### PR TITLE
examples: unify example apps across all example platforms

### PR DESCRIPTION
**Pull request by jwhui**

### CODE REVIEW

**Comment by turon** [comment]
I like the new structure overall!  Will review in detail.

File: examples/Makefile-cc2538 [hunk 1]
```diff
@@ -119,7 +116,7 @@ define configure-arch
 $(ECHO) "  CONFIG   $(1)-$(TargetTuple)..."
 (cd $(BuildPath)/$(1)-$(TargetTuple) && $(AbsTopSourceDir)/configure \
 INSTALL="$(INSTALL) $(INSTALLFLAGS)" \
-CPP="$(CPP)" CC="$(CC)" CXX="$(CXX)" OBJC="$(OBJC)" OBJCXX="$(OBJCXX)" AR="$(AR)" RANLIB="$(RANLIB)" NM="$(NM)" STRIP="$(STRIP)" CPPFLAGS="$(CPPFLAGS)" CFLAGS="$(CFLAGS)" CXXFLAGS="$(CXXFLAGS)" LDFLAGS="$(LDFLAGS)" LIBS="$(LIBS)" \
+CPP="$(CPP)" CC="$(CC)" CXX="$(CXX)" OBJC="$(OBJC)" OBJCXX="$(OBJCXX)" AR="$(AR)" RANLIB="$(RANLIB)" NM="$(NM)" STRIP="$(STRIP)" CPPFLAGS="$(CPPFLAGS)" CFLAGS="$(CFLAGS)" CXXFLAGS="$(CXXFLAGS)" LDFLAGS="$(LDFLAGS)" \
```
Comment (turon) [inline examples/Makefile-cc2538]: Noticed that `LIBS="$(LIBS)"` was removed -- is that ok?

File: examples/apps/cli/Makefile.am [hunk 2]
```diff
@@ -28,28 +28,36 @@
...
-ot_cli_LDADD                                                      = \
-    $(top_builddir)/src/core/libopenthread.a                        \
-    $(top_builddir)/src/cli/libopenthread-cli.a                     \
-    $(top_builddir)/examples/cc2538/platform/libopenthread-cc2538.a \
-    $(top_builddir)/third_party/mbedtls/libmbedcrypto.a             \
+ot_cli_LDADD                                                           = \
+    $(top_builddir)/src/core/libopenthread.a                             \
+    $(top_builddir)/src/cli/libopenthread-cli.a                          \
+    $(top_builddir)/third_party/mbedtls/libmbedcrypto.a                  \
+    -lstdc++                                                             \
```
Comment (turon) [inline examples/apps/cli/Makefile.am]: Do we need a dependency on `libstdc++`?

File: examples/apps/ncp/Makefile.am [hunk 3]
```diff
@@ -28,25 +28,36 @@
...
-    $(top_builddir)/src/ncp/libopenthread-ncp.a                   \
-    $(top_builddir)/src/core/libopenthread.a                      \
-    $(top_builddir)/examples/posix/platform/libopenthread-posix.a \
-    $(top_builddir)/third_party/mbedtls/libmbedcrypto.a           \
-    -lstdc++                                                      \
+ot_ncp_LDADD                                                           = \
+    $(top_builddir)/src/core/libopenthread.a                             \
+    $(top_builddir)/src/ncp/libopenthread-ncp.a                          \
+    $(top_builddir)/third_party/mbedtls/libmbedcrypto.a                  \
+    -lstdc++                                                             \
```
Comment (turon) [inline examples/apps/ncp/Makefile.am]: Same question as above: is -lstdc++ needed?

File: examples/Makefile-cc2538 (hunk 1 above)
Comment (jwhui) [inline examples/Makefile-cc2538]: Yes, that was intentional.  Nothing in OpenThread uses `LIBS`.

File: examples/platforms/posix/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+1. Spawn the process:
+
+```bash
+$ cd openthread/examples/posix/app/cli
+$ ./ot-cli 1
+```
+
+2. Type `help` for list of commands.
+
```
Comment (turon) [inline examples/platforms/posix/README.md]: Would it make sense to keep the common cli docs (command set) with the app, and make this README focus on specifics of the platform?

File: examples/apps/cli/Makefile.am (hunk 2 above)
Comment (jwhui) [inline examples/apps/cli/Makefile.am]: If you compile without, you get the following errors:
```bash
Undefined symbols for architecture x86_64:
  "vtable for __cxxabiv1::__class_type_info", referenced from:
      typeinfo for Thread::Cli::Server in libopenthread-cli.a(libopenthread_cli_a-cli_serial.o)
      typeinfo for Thread::Ip6::Netif in libopenthread.a(libopenthread_a-thread_netif.o)
      typeinfo for Thread::Ip6::Netif in libopenthread.a(libopenthread_a-netif.o)
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for __cxxabiv1::__si_class_type_info", referenced from:
      typeinfo for Thread::Cli::Serial in libopenthread-cli.a(libopenthread_cli_a-cli_serial.o)
      typeinfo for Thread::ThreadNetif in libopenthread.a(libopenthread_a-thread_netif.o)
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "___cxa_pure_virtual", referenced from:
      vtable for Thread::Ip6::Netif in libopenthread.a(libopenthread_a-netif.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: examples/apps/ncp/Makefile.am (hunk 3 above)
Comment (jwhui) [inline examples/apps/ncp/Makefile.am]: See above.

File: examples/platforms/cc2538/serial.c [hunk 5]
```diff
@@ -37,6 +37,7 @@
 #include <openthread-types.h>
 #include <common/code_utils.hpp>
 #include <platform/serial.h>
+#include "platform-cc2538.h"
 #include "cc2538-reg.h"
```
Comment (turon) [inline examples/platforms/cc2538/serial.c]: move the cc2538-reg include to within platform-cc2538.h ?

File: examples/platforms/posix/README.md (hunk 4 above)
Comment (jwhui) [inline examples/platforms/posix/README.md]: That's what I did here.

File: examples/apps/cli/Makefile.am (hunk 2 above)
Comment (turon) [inline examples/apps/cli/Makefile.am]: Perhaps using a build flag to allow a port to specify what stdc++ lib it wants to use would be good.  On many embedded ports, these __cxa calls would be stubbed out or handled by libnano.

File: examples/platforms/cc2538/serial.c (hunk 5 above)
Comment (jwhui) [inline examples/platforms/cc2538/serial.c]: Makes sense.  I will update.

**Comment by turon** [comment]
Just want to make clear I don't consider any of my comments blocking.  Approve!

**Pull request merged**